Add section to Snowflake tutorial on multiple records (#17615)
cockroachdb/docs: src/current/cockroachcloud/stream-changefeed-to-snowflake-aws.md

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    CREATE DATABASE cdc_test;
    ~~~

1. Set it as the default:

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    SET DATABASE = cdc_test;
    ~~~

## Step 4. Create tables

Before you can start a changefeed, you need to create at least one table for the changefeed to target. The targeted table's rows are referred to as the "watched rows".

Create a table called `order_alerts` to target:

{% include_cached copy-clipboard.html %}
~~~ sql
CREATE TABLE order_alerts (

>     id   INT PRIMARY KEY,

Reviewer: ```suggestion
    id INT PRIMARY KEY,
```

    name STRING
);
~~~

## Step 5. Create an S3 bucket in the AWS Console

Every change to a watched row is emitted as a record in a configurable format (i.e., `JSON` for cloud storage sinks). To configure an AWS S3 bucket as the cloud storage sink:

1. Log in to your [AWS S3 Console](https://s3.console.aws.amazon.com/).

1. Create an S3 bucket where streaming updates from the watched tables will be collected.

>     You will need the name of the S3 bucket when you [create your changefeed](#step-6-create-an-enterprise-changefeed). Ensure you have a set of IAM credentials with write access on the S3 bucket that you will use during [changefeed setup](#step-6-create-an-enterprise-changefeed).

Reviewer: can call out this information in a note

Author: I think this is really part of the workflow — where possible I do not want to overuse notes/warnings/tips as it then diminishes there usefulness throughout the docs.

Reviewer: Okay, makes sense to me!

## Step 6. Create an enterprise changefeed

> Back in the built-in SQL shell, [create an enterprise changefeed](../{{site.current_cloud_version}}/create-changefeed.html). Replace the placeholders with your AWS access key ID and AWS secret access key:

Author: I'm not quite sure what you mean here, but there are no sub-steps here because there is only one command the user is running.

> Back in the built-in SQL shell, [create an enterprise changefeed](../{{site.current_cloud_version}}/create-changefeed.html). Replace the placeholders with your AWS access key ID and AWS secret access key:

Reviewer: That makes sense - while two events happen here, only one is a command so leaving as is works

{% include_cached copy-clipboard.html %}
~~~ sql
CREATE CHANGEFEED FOR TABLE order_alerts
    INTO 's3://{bucket name}?AWS_ACCESS_KEY_ID={access key ID}&AWS_SECRET_ACCESS_KEY={secret access key}'
    WITH
        updated;
~~~

> Refer to the [Cloud Storage Authentication](../{{site.versions["stable"]}}/cloud-storage-authentication.html) page for more detail on authenticating to Amazon S3 and other cloud providers.

Reviewer: i think this would make sense to be at the end of step 6, following the paragraph on the changefeed job ID

Author: Done

1. Log in to Snowflake as a user with [read and write access](https://docs.snowflake.net/manuals/user-guide/security-access-control-overview.html) to a cluster.

> 1. Navigate to the **Worksheets** and select a worksheet.

Reviewer: ```suggestion
1. Navigate to the **Worksheets** page and select a worksheet.
```

Reviewer: Snowflake calls it a page here https://docs.snowflake.com/en/user-guide/ui-worksheet but we could also say Navigate to **Worksheets**

Author: I'm not sure why we would want to change this when it is called the Worksheets page?

Reviewer: apologies, I meant either calling it 1. the **Worksheets** page or 2. Navigate to **Worksheets** instead of saying "Navigate to the **Worksheets**..."

    ~~~

    {{site.data.alerts.callout_info}}
    Auto-ingest in Snowflake works with [AWS](https://docs.snowflake.com/user-guide/data-load-snowpipe-auto-s3.html), [Azure](https://docs.snowflake.net/manuals/user-guide/data-load-snowpipe-auto-azure.html), and [Google Cloud Storage](https://docs.snowflake.com/user-guide/data-load-snowpipe-auto-gcs.html).
    {{site.data.alerts.end}}

1. In the worksheet, view the Snowpipe:

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    SHOW PIPES;
    ~~~

1. Copy the **ARN** of the SQS queue for your stage, which displays in the **notification_channel** column. You will use this information to [configure the S3 bucket](#step-9-configure-the-s3-bucket).

## Step 9. Configure the S3 bucket

1. [Navigate back to your S3 bucket](https://s3.console.aws.amazon.com/).

1. Configure an event notification for the S3 bucket. In the **Properties** tab, click **Create event notification**. Use the following parameters:
    - **Event Name:** Name of the event notification (e.g., Auto-ingest Snowflake).
    - **Event types:** Select the **All object create events**.
    - **Destination:** Select **SQS Queue**.

>     - **Specify SQS queue:** Select **Enter SQS queue ARN** from the drop-down.
>     - **SQS queue ARN:** Paste the SQS queue name from the `SHOW PIPES` output (from [Step 8](#step-8-configure-snowflake)).

Reviewer: ```suggestion
    - **Specify SQS queue:** Select **Enter SQS queue ARN** from the drop-down.
    - **SQS queue ARN:** Paste the SQS queue name from the `SHOW PIPES` output (from [Step 8](#step-8-configure-snowflake)).
```

nit: nest these two points within **Destination** as they are parameters within the Destination configuration

Author: I don't think this quite makes a difference here. They're actually in a separate section in the UI. And as we're just directing the user through each click, it kind matters more what is happening in our instructions.

Reviewer: Got it. When looking at the S3 page, I see that it is under the Destination section but the following flow already guides a user down the page and works.

Your changefeed is now streaming to Snowflake.

## Remove multiple records from Snowflake target tables

In some cases, you may need to remove multiple records for the same primary key caused by updates to a row. Although it is not possible to remove duplicates with Snowpipe before they arrive at the target table, Snowflake offers some features that you can use to automate removing multiple records from the table downstream.

The following points outline two potential workarounds. For detailed instructions on each feature, we recommend reading the linked Snowflake documentation:

- Use Snowflake [streams](https://docs.snowflake.com/en/user-guide/streams-intro) and [tasks](https://docs.snowflake.com/en/user-guide/tasks-intro) to write into a new de-duplicated table.
    - Set up a stream that will watch for changes on the table that Snowpipe uploads into. The stream will create a snapshot of the changes at a specific point in time.

>     - Create the new table in Snowflake that will hold the de-duplicated entries using the stream's `METADATA$ACTION` column.

Reviewer: ```suggestion
    - Create a new table in Snowflake that will hold the de-duplicated entries using the stream's `METADATA$ACTION` column.
```

Author: "a" is fine here I think. Since we're creating "a" new table not referring back to "the" table we have already created.

    - Create a task to run a SQL statement that will pull data from the stream and merge it into a new table for the "unique" entries. You can set this task to run when there are new records in the stream and by a cron job schedule.

>     Refer to the [stream creation examples in Snowflake's documentation](https://docs.snowflake.com/en/user-guide/streams-examples#basic-example).

Reviewer: ```suggestion
    Refer to [Snowflake's examples on streams](https://docs.snowflake.com/en/user-guide/streams-examples#basic-example).
```
